feat(RV64): implement dynamic memory/cpu decider

PeterWrighten opened the pull request:
## Description
In order to close #463 .

Register and implement Interrupt Controller and Timer.

FDT Based Dynamic Decider is implemented as the following design.

1. Created FDT(Flattened Device Tree) parser for dynamic memory detection from device tree, with
fallback mechanisms. (Using [fdt](https://docs.rs/fdt/latest/fdt/) crate)

2. Dynamic Heap Size and Start Decider implementation in vm.rs
3. Dynamic CPU Counts implementation.

## How was this PR tested?

I tested this heap decider with QEMU memory size 512M, 1G,  and 2G. Test results are shown as the following.
Both heap and VM are initialized successfully by using Dynamic Heap Decider.

```shell
// 512M Memory Size

qemu-system-riscv64 -machine virt -bios none -kernel target/riscv64gc-unknown-none-elf/release/awkernel -m 512M -nographic -smp 4 -monitor telnet::5556,server,nowait

=== AWkernel RV64 Starting ===
Console port initialized
UART driver initialized
Setting up heap allocation...
Memory detected from boot DTB pointer at 0x00009fe09fe00000
Detected RAM size: 0x0000000020000000
Using dynamic heap configuration\r\nHeap allocation complete\r\nInitializing page allocator...
Initializing kernel space...
Memory detected from boot DTB pointer at 0x00009fe09fe00000
Detected RAM size: 0x0000000020000000
Mapping kernel text section...
Mapping kernel rodata section...
Mapping kernel data section...
Mapping kernel bss section...
Mapping physical memory...
  from ekernel: 0x800d7000
  to MEMORY_END: 0xa0000000 (dynamic)
Memory detected from boot DTB pointer at 0x00009fe09fe00000
Detected RAM size: 0x0000000020000000
Dynamic heap calculation:
  Kernel end: 0x800d7000
  Heap start: 0x800d7000
  Memory end: 0xa0000000
  Max heap size: 0x1ff29000
Activating kernel space...
Virtual memory system initialized
Initializing RISC-V interrupt controller...\r\nInitializing RISC-V timer...\r\nTimer and interrupt controller initialized\r\n[            3 INFO] AWkernel RV64 primary CPU initialization complete
[            5 INFO] CPU#0 is starting.
[            6 INFO] CPU#2 is starting.
[            6 INFO] CPU#3 is starting.
[            7 INFO] CPU#1 is starting.
[            8 INFO] interrupt::INTERRUPT_CONTROLLER has been initialized.
[            9 INFO] interrupt::PREEMPT_FN has been initialized.
[           10 INFO] timer::TIMER has been initialized.
[           12 INFO] A local timer has been initialized.
```

```shell
// 1G Memory Size

qemu-system-riscv64 -machine virt -bios none -kernel target/riscv64gc-unknown-none-elf/release/awkernel -m 1G -nographic -smp 4 -monitor telnet::5556,server,nowait

=== AWkernel RV64 Starting ===
Console port initialized
UART driver initialized
Setting up heap allocation...
Memory detected from boot DTB pointer at 0x0000bfe0bfe00000
Detected RAM size: 0x0000000040000000
Using dynamic heap configuration\r\nHeap allocation complete\r\nInitializing page allocator...
Initializing kernel space...
Memory detected from boot DTB pointer at 0x0000bfe0bfe00000
Detected RAM size: 0x0000000040000000
Mapping kernel text section...
Mapping kernel rodata section...
Mapping kernel data section...
Mapping kernel bss section...
Mapping physical memory...
  from ekernel: 0x800d7000
  to MEMORY_END: 0xc0000000 (dynamic)
Memory detected from boot DTB pointer at 0x0000bfe0bfe00000
Detected RAM size: 0x0000000040000000
Dynamic heap calculation:
  Kernel end: 0x800d7000
  Heap start: 0x800d7000
  Memory end: 0xc0000000
  Max heap size: 0x3ff29000
Activating kernel space...
Virtual memory system initialized
Initializing RISC-V interrupt controller...\r\nInitializing RISC-V timer...\r\nTimer and interrupt controller initialized\r\n[            3 INFO] AWkernel RV64 primary CPU initialization complete
[            5 INFO] CPU#0 is starting.
[            6 INFO] CPU#3 is starting.
[            6 INFO] CPU#1 is starting.
[            7 INFO] CPU#2 is starting.
[            7 INFO] interrupt::INTERRUPT_CONTROLLER has been initialized.
[            8 INFO] interrupt::PREEMPT_FN has been initialized.
[            9 INFO] timer::TIMER has been initialized.
[           11 INFO] A local timer has been initialized.
```

```shell
// 2G Memory Size

qemu-system-riscv64 -machine virt -bios none -kernel target/riscv64gc-unknown-none-elf/release/awkernel -m 2G -nographic -smp 4 -monitor telnet::5556,server,nowait

=== AWkernel RV64 Starting ===
Console port initialized
UART driver initialized
Setting up heap allocation...
Memory detected from boot DTB pointer at 0x0000bfe0bfe00000
Detected RAM size: 0x0000000080000000
Using dynamic heap configuration\r\nHeap allocation complete\r\nInitializing page allocator...
Initializing kernel space...
Memory detected from boot DTB pointer at 0x0000bfe0bfe00000
Detected RAM size: 0x0000000080000000
Mapping kernel text section...
Mapping kernel rodata section...
Mapping kernel data section...
Mapping kernel bss section...
Mapping physical memory...
  from ekernel: 0x800d7000
  to MEMORY_END: 0x00000000 (dynamic)
Memory detected from boot DTB pointer at 0x0000bfe0bfe00000
Detected RAM size: 0x0000000080000000
Dynamic heap calculation:
  Kernel end: 0x800d7000
  Heap start: 0x800d7000
  Memory end: 0x00000000
  Max heap size: 0x7ff29000
Activating kernel space...
Virtual memory system initialized
Initializing RISC-V interrupt controller...\r\nInitializing RISC-V timer...\r\nTimer and interrupt controller initialized\r\n[            3 INFO] AWkernel RV64 primary CPU initialization complete
[            6 INFO] CPU#0 is starting.
[            6 INFO] CPU#3 is starting.
[            7 INFO] CPU#2 is starting.
[            8 INFO] CPU#1 is starting.
[            8 INFO] interrupt::INTERRUPT_CONTROLLER has been initialized.
[            9 INFO] interrupt::PREEMPT_FN has been initialized.
[           10 INFO] timer::TIMER has been initialized.
[           12 INFO] A local timer has been initialized.
```
After Implementation of Interrupt Controller and timer, the RV64 version kernel worked well as the following.

```shell
// All basic function including timer, interrupt controller and dynamic cpu counts worked well
make qemu-riscv64
qemu-system-riscv64 -machine virt -bios none -kernel target/riscv64gc-unknown-none-elf/release/awkernel -m 1G -nographic -smp 4 -monitor telnet::5556,server,nowait

=== AWkernel RV64 Starting ===
Console port initialized
UART driver initialized
Setting up heap allocation...
Memory detected from boot DTB pointer at 0x0000bfe0bfe00000
Detected RAM size: 0x0000000040000000
Using dynamic heap configuration
Heap allocation complete
Initializing page allocator...
Initializing kernel space...
Memory detected from boot DTB pointer at 0x0000bfe0bfe00000
Detected RAM size: 0x0000000040000000
Mapping kernel text section...
Mapping kernel rodata section...
Mapping kernel data section...
Mapping kernel bss section...
Mapping physical memory...
  from ekernel: 0x00000000800df000
  to MEMORY_END: 0x00000000c0000000 (dynamic)
Memory detected from boot DTB pointer at 0x0000bfe0bfe00000
Detected RAM size: 0x0000000040000000
Dynamic heap calculation:
  Kernel end: 0x00000000800df000
  Heap start: 0x00000000800df000
  Memory end: 0x00000000c0000000
  Max heap size: 0x000000003ff21000
Activating kernel space...
Virtual memory system initialized
Initializing RISC-V interrupt controller...
Initializing RISC-V timer...
Timer and interrupt controller initialized
[            1 INFO] AWkernel RV64 primary CPU initialization complete
Detected 00000004 CPUs from boot DTB
[            5 INFO] CPU#0 is starting.
[            6 INFO] CPU#1 is starting.
[            6 INFO] CPU#2 is starting.
[            7 INFO] CPU#3 is starting.
[            7 INFO] interrupt::INTERRUPT_CONTROLLER has been initialized.
[            8 INFO] interrupt::PREEMPT_FN has been initialized.
[            9 INFO] timer::TIMER has been initialized.
[           12 INFO] A local timer has been initialized.
[           15 INFO] Starting [Awkernel] network service.
```

PeterWrighten committed:
feat(RV64): implement dynamic memory/cpu decider, timer and interrupt

- Created FDT parser for dynamic memory/cpu detection from device tree,
  with fallback mechanisms
- Register and implement RISC-V timer
- Register and implement interrupt controller and IPI

PeterWrighten reviewed (commented):
[WIP] Some Debugging code Modification required

ytakano commented on awkernel_lib/src/arch/rv64.rs:
```diff
@@ -39,3 +58,128 @@ pub fn translate_kernel_address(vpn: address::VirtPageNum) -> Option<page_table:
+        }
+
+        // Fall back to default
+        unsafe_puts("Using default CPU count\r\n");
```
Why don't you stop booting, instead of using the default CPU count?

PeterWrighten commented on awkernel_lib/src/arch/rv64/vm.rs:
```diff
@@ -153,6 +154,17 @@ impl MemorySet {
         }
     }
 
+    #[allow(dead_code)]
```
Because the CI/CD has cargo clippy_rv64 test, this is for passing it.
```shell
 cargo clippy_rv64 -- --deny warnings
```

PeterWrighten committed:
refactor(RV64): refactor Dynamic Decider's memory end detector for explicit implementation

ytakano commented:
> I consider the following logics would make frame allocator exhausted, but seems a balanced solution.
> 
> ```rust
> while let Some(frame) in frame_allocator.alloc() {
>          page_table.map_to(heap_virt_addr_start + i * PAGE_SIZE, frame.phy_addr, &mut frame_allocator);
>          i += 1;
> }
> ```

Yes.
By using this algorithm, all memory frame will be used by either the heap memory or the page table.

PeterWrighten committed:
fix(RV64): eliminate frame allocator / heap overlap

Both the frame allocator and TLSF heap previously started at ekernel.
find_pte_create() overwrote TLSF metadata when handing out page-table
node frames, causing heap corruption after kernel space initialization.

Split physical memory into non-overlapping regions:
  [ekernel, ekernel+N)    — frame allocator (PT nodes only)
  [ekernel+N, memory_end) — heap (backup + primary)

N = 10% of available RAM, clamped to [4 MiB, 256 MiB].
Removes hardcoded MEMORY_END; init_page_allocator and init_kernel_space
now take explicit physical bounds.

PeterWrighten commented:
## FYI
This PR is currently including both #675 and #676 , not split into `dynamic heap decider only`. 

As mentioned in [this comment](https://github.com/tier4/awkernel/pull/676#issuecomment-4311535533),  we should consider whether it is necessary to still work on PR seperation or just keep this PR for whole features.

PeterWrighten committed:
feat(asm): fix and refine asm error and interrupt controller

PeterWrighten commented on kernel/src/arch/rv64/kernel_main.rs:
```diff
@@ -82,39 +122,195 @@ unsafe fn primary_hart(hartid: usize) {
+    // 2. Heap over [heap_start, memory_end) — no overlap with frame allocator.
+    uart_debug_puts("Initializing heap...\r\n");
+    let backup_size = BACKUP_HEAP_SIZE;
+    let primary_size = memory_end - heap_start - backup_size;
```
add gurantee for sufficient RAM.

PeterWrighten committed:
fix(RV64): address review findings on interrupt dispatch and memory mapping

- Dispatch local interrupts (MTIP timer, MSIP IPIs) via a per-hart
  LOCAL_PENDING bitmask instead of the PLIC claim path, which never
  reports them; send_ipi now records the logical IRQ (preempt/wakeup)
- Move TIMER_IRQ from PLIC source 5 to synthetic ID 63 and skip PLIC
  programming for local IRQs
- Drain all pending PLIC sources by looping claim/complete in
  pending_irqs() instead of claiming at most one
- Handle machine external interrupts (mcause=11) in boot.S and enable
  MEIE; initialize primary hart PLIC threshold via init_primary()
- Identity-map the page-table-frame region [pt_start, heap_start) in
  init_kernel_space() since PT frames are dereferenced by phys address
- Panic if the 4 MiB page-table-frame floor exceeds available RAM
- Update stale heap-mapping comment in vm.rs

ytakano commented on awkernel_lib/src/interrupt.rs:
```diff
@@ -403,6 +403,57 @@ pub fn handle_preemption() {
+/// Handle all pending interrupt requests for RISC-V.
+/// This is called from the M-mode interrupt handler.
+#[cfg(any(target_arch = "riscv32", target_arch = "riscv64"))]
+pub fn handle_irqs(is_task: bool) {
```
Is `handle_irqs()` same as AArch64?
Consider remove this definition if possible and add the `target_arch` of RISCV to the AArch64's `handle_irqs()` defined above.

ytakano commented on kernel/src/arch/rv64/kernel_main.rs:
```diff
@@ -82,39 +140,215 @@ unsafe fn primary_hart(hartid: usize) {
+    // 1. Frame allocator over [pt_start, pt_end) — no heap needed yet.
+    uart_debug_puts("Initializing frame allocator (PT frames: 0x");
+    awkernel_lib::console::unsafe_print_hex_u64(pt_frames_size as u64);
+    uart_debug_puts(" bytes)\r\n");
```
`init_memory()` gives the frame allocator the region starting immediately after `ekernel`, but `boot.S` still places the initial per-hart stacks at fixed addresses above `0x80800000`. With the tested 512 MiB configuration, the page-table-frame region can extend across those stack addresses, so later page-table allocations may reuse memory that is still being used as CPU stacks. Please reserve/exclude the initial stack area from the frame allocator and heap layout, or move the stacks into a region that is explicitly reserved before initializing the allocator.

ytakano commented on kernel/src/arch/rv64/interrupt_controller.rs:
```diff
@@ -0,0 +1,266 @@
+            }
+            pending.push(claimed as u16);
+            // Complete the interrupt (write back the claim)
+            unsafe { write_volatile(claim_reg, claimed) };
```
`pending_irqs()` completes each PLIC interrupt immediately after claiming it, before the registered handler runs in `awkernel_lib::interrupt::handle_irqs()`. For level-triggered devices, completing before the handler clears the device-side interrupt condition can cause the same source to be re-pended or handled out of order. Please move PLIC completion to after the handler has run, or restructure the RV64 external interrupt path so claim, handler dispatch, and complete happen in the correct order.

ytakano reviewed (commented):
Adding inline comments for the remaining two issues.

PeterWrighten committed:
fix(RV64): dedup handle_irqs, reserve boot stacks, defer PLIC completion
